cmd/snap-bootstrap/triggerwatch/evdev.go

```diff
@@ -0,0 +1,242 @@
+		"KEY_2":   evdev.KEY_2,
+		"KEY_3":   evdev.KEY_3,
+		"KEY_4":   evdev.KEY_4,
+		"KEY_5":   evdev.KEY_5,
+		"KEY_6":   evdev.KEY_6,
+		"KEY_7":   evdev.KEY_7,
+		"KEY_8":   evdev.KEY_8,
+		"KEY_9":   evdev.KEY_9,
+		"KEY_0":   evdev.KEY_0,
+	}
+	evKeyCapability = evdev.CapabilityType{Type: evdev.EV_KEY, Name: "EV_KEY"}
+
+	// hold time needed to trigger the event
+	holdToTrigger = 2 * time.Second
+)
+
+func init() {
+	trigger = &evdevInput{}
+}
+
+type evdevKeyboardInputDevice struct {
+	keyCode uint16
+	dev     *evdev.InputDevice
+}
+
+func (e *evdevKeyboardInputDevice) probeKeyState() (bool, error) {
+	// XXX: evdev defines EVIOCGKEY using MAX_NAME_SIZE which is larger than
+	// what is needed to store the key bitmap with KEY_MAX bits, but we need
+	// to play along since the value is already encoded
+	keyBitmap := new([evdev.MAX_NAME_SIZE]byte)
+
+	// obtain the large bitmap with all key states
+	// https://elixir.bootlin.com/linux/v5.5.5/source/drivers/input/evdev.c#L1163
+	_, _, err := syscall.RawSyscall(syscall.SYS_IOCTL, e.dev.File.Fd(), uintptr(evdev.EVIOCGKEY), uintptr(unsafe.Pointer(keyBitmap)))
+	if err != 0 {
+		return false, err
+	}
+	byteIdx := e.keyCode / 8
+	keyMask := byte(1 << (e.keyCode % 8))
+	isDown := keyBitmap[byteIdx]&keyMask != 0
+	return isDown, nil
+}
+
+func (e *evdevKeyboardInputDevice) WaitForTrigger(ch chan keyEvent) {
+	logger.Noticef("%s: starting wait", e)
+
+	// XXX: do not mess with setting the key repeat rate, as it's cumbersome
+	// and golang-evdev SetRepeatRate() parameter order is actually reversed
+	// wrt. what the kernel does. The evdev interprets EVIOCSREP arguments
+	// as (delay, repeat)
+	// https://elixir.bootlin.com/linux/latest/source/drivers/input/evdev.c#L1072
+	// but the wrapper is passing is passing (repeat, delay)
+	// https://github.com/gvalkov/golang-evdev/blob/287e62b94bcb850ab42e711bd74b2875da83af2c/device.go#L226-L230
+
+	holdState, err := e.probeKeyState()
+	if err != nil {
+		ch <- keyEvent{Err: fmt.Errorf("cannot obtain initial key state: %v", err), Dev: e}
+	}
+	if holdState {
+		// looks like the key is pressed initially, we don't know when
+		// that happened, but prentend got pressed now
+		logger.Noticef("%s: key is already down", e)
+	}
+
+	type evdevEvent struct {
+		kev *evdev.KeyEvent
+		err error
+	}
+
+	// buffer large enough to collect some events
+	evChan := make(chan evdevEvent, 10)
+
+	monitorKey := func() {
+		for {
+			ies, err := e.dev.Read()
+			if err != nil {
+				evChan <- evdevEvent{err: err}
+				break
+			}
+			for _, ie := range ies {
+				if ie.Type != evdev.EV_KEY || ie.Code != e.keyCode {
+					continue
+				}
+				kev := evdev.NewKeyEvent(&ie)
+				evChan <- evdevEvent{kev: kev}
+			}
+		}
+		close(evChan)
+	}
+
+	go monitorKey()
+
+	holdTimer := time.NewTimer(holdToTrigger)
+	// no sense to keep it running later either
+	defer holdTimer.Stop()
+
+	if !holdState {
+		// key isn't held yet, stop the timer
+		holdTimer.Stop()
+	}
+Loop:
```

Comment:
maybe good to add a comment before Loop, like:

// invariant:  tholdTimer is running iff holdState is true, otherwise is stopped